File: MATTHEW/leetcode/AIcoding/4.GPT-4o/java_solutions/2516-take-k-of-each-character-from-left-and-right.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public String takeCharacters(String s, int k) {
        Map<Character, Integer> count = new HashMap<>();
        int n = s.length();
        int left = 0, right = 0, maxLength = 0;

        while (right < n) {
            count.put(s.charAt(right), count.getOrDefault(s.charAt(right), 0) + 1);
            while (count.size() > k) {
                count.put(s.charAt(left), count.get(s.charAt(left)) - 1);
                if (count.get(s.charAt(left)) == 0) {
                    count.remove(s.charAt(left));
                }
                left++;
            }
            maxLength = Math.max(maxLength, right - left + 1);
            right++;
        }

        return s.substring(0, maxLength);
    }
}

File: leetcode/AIcoding/2.Gemini-2.0/javascript_solutions/2516-take-k-of-each-character-from-left-and-right.js
var takeCharacters = function(s, k) {
    let n = s.length;
    let countA = 0, countB = 0, countC = 0;
    for (let char of s) {
        if (char === 'a') countA++;
        else if (char === 'b') countB++;
        else countC++;
    }

    if (countA < k || countB < k || countC < k) return -1;

    let left = 0;
    let maxLen = 0;
    let a = 0, b = 0, c = 0;

    for (let right = 0; right < n; right++) {
        if (s[right] === 'a') a++;
        else if (s[right] === 'b') b++;
        else c++;

        while (a > countA - k || b > countB - k || c > countC - k) {
            if (s[left] === 'a') a--;
            else if (s[left] === 'b') b--;
            else c--;
            left++;
        }

        maxLen = Math.max(maxLen, right - left + 1);
    }

    return n - maxLen;
};